proxy/http/HttpTransact.h

```diff
@@ -944,7 +944,22 @@ class HttpTransact
     MgmtInt
     configured_connect_attempts_max_retries() const
     {
-      return txn_conf->connect_attempts_max_retries;
+      if (dns_info.looking_up != ResolveInfo::PARENT_PROXY) {
+        return txn_conf->connect_attempts_max_retries;
+      }
+      // For parent proxy, return the maximum attempt count for the current
+      // parent intead of the global max retries for the whole parent group.
+      // The max attempt count for the current parent is calculated by rounding
+      // the current attempt up to next multiple of ppca.
+      auto ppca                    = txn_conf->per_parent_connect_attempts;
+      auto cur_attempts            = current.attempts.get();
+      auto cur_parent_max_attempts = ((cur_attempts + ppca - 1) / ppca) * ppca;
+      if (cur_attempts == cur_parent_max_attempts) {
+        // If the current attempt is already a multiple of ppca, get the next
+        // multiple.
+        cur_parent_max_attempts += ppca;
+      }
```

Comment:
I don't think you want this `if` statement.  If `ppca` is 10 and `cur_attempts` is 20, that means you're on the last try to connect to the second origin in the next hop protection group.  If there's a non-retryable error,  you don' t want to raise the number of attempts to 30.  That would mean you'd never try the last origin in the next hop protection group.

Reply:
Without this, when `cur_attempts` is already a multiple of `ppca` when calling this function, the calculation would always return the same `cur_attempts`, which leads to an assertion error in `increment()` down the line.

I do see your point though and we don't want to skip parents. Question: when `cur_attempts` is 20, is it the last try for the second origin, or is it the first try for the last origin?

Reply:
This call: https://github.com/apache/trafficserver/blob/6e8933f35ebf5a50211e3b06aa52749ac0ec3c3b/proxy/http/HttpTransact.cc#L3646 can't cause an assert.  `configured_connect_attempts_max_retries()` can't be more than `txn_conf->parent_connect_attempts`.  The test in the `if` statement at line 3644 ensures that an assert is impossible if execution reaches line 3644.

The only other call to the increment function is https://github.com/apache/trafficserver/blob/6e8933f35ebf5a50211e3b06aa52749ac0ec3c3b/proxy/http/HttpTransact.cc#L3865.  Not clear under what circumstances this call is reached.  The very large HttpTransact and HttpSM classes have little internal scoping or even encapsulation.  It could takes days or weeks to analyze possible code paths.  As I previously said, it's very hard to work on a source bases like this without taking leaps of faith and making reasonable assumptions.  So I suggest making the reasonable assumption that a retry will be blocked if the number of retries is already at the maximum.  To back that up, you could write long Au test(s) exercising parent.config (as I did for strategies.yaml).  But if parent.config is soon to be deprecated, it's questionable if the effort involved is worth it.

Reply:
Right. But the same logic applies when calling `maximize()` with the `configured_connect_attempts_max_retries()`:
```
current.attempts : 0
per_parent_connect_attempts: 2
parent_proxy.total_connect_attempts: 100
```
the `configured_connect_attempts_max_retries()` returns 0, which is passed into `maximize()`. As a result, `current.attempts` is unchanged after the call, while what we want is to "maximize" the attempt count so that the current parent is skipped. There seems to be an issue here too. Do you have any suggestions on this?

Reply:
From looking at the code, I thought the attempt numbers are 1 to N.  That is the number is 1 during the first attempt, not zero.  So the attempt number would never be zero when `maximize()` is called.

Reply:
OK, I see now, "retries" doesn't include the first try, so max tries = max retries + 1.

Reply:
How about?
```
    MgmtInt
    configured_connect_attempts_max_retries() const
    {
      if (parent_params != nullptr) {
        auto ppca = txn_conf->per_parent_connect_attempts;
        auto tries = current.attempts.get() + 1;
        // Round g up to next multiple of ppca.
        auto ru = ((tries + ppca - 1) / ppca) * ppca;
        return std::min(ru, txn_conf->parent_connect_attempts) - 1;
      }
      return txn_conf->connect_attempts_max_retries;
    }
```
But this would also require removing the parameter and assert in `CurrentInfo::Attempts::increment()`, 
and changing line 3644 in HttpTransact.cc to:
```
    if (s->current.attempts.get() < (s->txn_conf->parent_connect_attempts - 1)) {
```
and changing line 3649 in HttpTransact.cc to:
```
      if (s->current.attempts.get() % s->txn_conf->per_parent_connect_attempts != 0) {
```

Reply:
Yep after some more digging and thoughts, I believe this is correct. I reach out to Susan and she confirms that the current `HttpTransact::_CurrentInfo.attempt` is really the retry attempt. Based on this understanding, what you suggested here makes a lot of sense and I think we are good to go.